src/System.Text.Json/src/System/Text/Json/Serialization/PooledBufferWriter.cs

```diff
@@ -98,18 +101,30 @@ public void Advance(int count)
             _index += count;
         }
 
-        public Memory<T> GetMemory(int sizeHint = 0)
+        public Memory<byte> GetMemory(int sizeHint = 0)
         {
             CheckAndResizeBuffer(sizeHint);
             return _rentedBuffer.AsMemory(_index);
         }
 
-        public Span<T> GetSpan(int sizeHint = 0)
+        public Span<byte> GetSpan(int sizeHint = 0)
         {
             CheckAndResizeBuffer(sizeHint);
             return _rentedBuffer.AsSpan(_index);
         }
 
+#if BUILDING_INBOX_LIBRARY
+        internal ValueTask WriteToAsync(Stream destination, CancellationToken cancellationToken)
+        {
+            return destination.WriteAsync(WrittenMemory, cancellationToken);
+        }
+#else
+        internal Task WriteToAsync(Stream destination, CancellationToken cancellationToken)
+        {
+            return destination.WriteAsync(_rentedBuffer, 0, _index, cancellationToken);
```

Comment:
I am not sure about our guidelines for passing in a pooled array to a stream passed in by the caller, which is partially why the initial `ToArray()` existed.

Reply:
cc @bartonjs, @GrabYourPitchforks

Reply:
> I am not sure about our guidelines for passing in a pooled array to a stream passed in by the caller

That's happening even with WrittenMemory, since the destination stream can fish out the underlying array via MemoryMarshal.TryGetArray.

We pass such arrays around in many other places.

Reply:
It is OK as long as we control the ownership and lifecycle of the buffer. In this case, the buffer is private and the type itself encapsulates all the logic behind renting/releasing/resing of the pooled buffer.

Also, as Stephen has pointed that's already happening with WrittenMemory (but the array is wrapped with ROM).

Reply:
> It is OK as long as we control the ownership and lifecycle of the buffer

Yes and no.  We've deemed it ok, but the issue Ahson is pointing out is that we're handing off a pooled buffer to arbitrary code we don't necessarily own.  It could take that array, stash it away and use it later, and then we're off to the races.

Reply:
> It could take that array, stash it away and use it later, and then we're off to the races

I would never think about writing a `Stream` that does that but now I understand the point Ahson made. 👍

Reply:
>  since the destination stream can fish out the underlying array via MemoryMarshal.TryGetArray.

Yes, except the user then had to use "unsafe" code in that case.

> We pass such arrays around in many other places.

Sounds good to me. I am fine with following precedence here.